### PR TITLE
[WIP] ZCurve indexing

### PR DESCRIPTION
**Pull request by achulkov2**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

For changelog. Remove if this is non-significant change.

Category (leave one):
- New Feature

Short description (up to few sentences):

Implements basic indexing with space-filling curves. Adds support for invertible functions in primary key. 

Detailed description (optional):

...

### CODE REVIEW

**Comment by javisantana** [comment]
could you provide a basic example?

**Comment by achulkov2** [comment]
> could you provide a basic example?
For example, if you have two columns X and Y with roughly the same number of SELECT queries with conditions on X and Y, then perhaps it is more efficient to order them by some special function f(X, Y), rather than using (X, Y) or (Y, X) in the key. In this case space-filling curves are used as the aforementioned function. As it currently stands, one would simply use ZCurve(X, Y) in the primary key for this to work. This will also work with more than two columns.

**Comment by javisantana** [comment]
is it like:

```
create table ( x Float32, y Float32, ....) ENGINE=MergeTree partition by (ZCurve(x, y)) order by ZCurve(x, y)
```

how would you query a range (for example, a bounding box in 2D)

**Comment by achulkov2** [comment]
> is it like:
> 
> ```
> create table ( x Float32, y Float32, ....) ENGINE=MergeTree partition by (ZCurve(x, y)) order by ZCurve(x, y)
> ```
> how would you query a range (for example, a bounding box in 2D)

I am not sure that partitioning by ZCurve(x, y) would be a very good idea because the function is  injective, so it can assume rather a lot of values. Aside from that it is exactly like you wrote. 
Range queries don't change: you would do something like SELECT ..... WHERE x > 0 and y > 0
It will automatically analyse the expression in the query and in the primary key and only choose blocks that may fulfil the query conditions. This will work in general with any function for which inevitability is implemented.

**Comment by javisantana** [comment]
ok, got it, thanks

About the partitioning I was thinking on creating partitions by geographical area but it'd more something like

```
ZCurve(toInt32(x/K), toInt32(y/K))
```

**Comment by alexey-milovidov** [comment]
> About the partitioning I was thinking on creating partitions by geographical area but it'd more something like ZCurve(toInt32(x/K), toInt32(y/K))

Partitions are independent. If you partition by `ZCurve(x, y)` it is the same as if you partition by tuple `(x, y)`. Moreover, partitions are not a replacement of primary key: if you have a primary key by `ZCurve(toInt32(x/K), toInt32(y/K))` you don't have to partition your data, because index will work fine.

**Comment by javisantana** [comment]
> If you partition by `ZCurve(x, y)` it is the same as if you partition by tuple `(x, y)`

I don't exactly know how CH sorts the data when the primary key is a tuple (I read it first order by x and then by y for each x). In the geospatial work the access pattern is usually different so if the curve function is able to keep close things (in the real world) closer in the disk is much better. 

For example: X, Y = (10, 10) and X2, Y2=(11, 11) which are close in the 2d space will be closer if you have a zcurve that make them closer (like Hilbert or quadkey). Not sure what pattern this ZCurve is actually implementing tho.

Here you have a visualization for quadkey, it shows the ratio between the euclidean distance and the quadkey distance to the center (darker is worse)

captura de pantalla 2019-03-01 a las 14 03 47

**Comment by alexey-milovidov** [comment]
> I don't exactly know how CH sorts the data when the primary key is a tuple (I read it first order by x and then by y for each x). In the geospatial work the access pattern is usually different so if the curve function is able to keep close things (in the real world) closer in the disk is much better.

> For example: X, Y = (10, 10) and X2, Y2=(11, 11) which are close in the 2d space will be closer if you have a zcurve that make them closer (like Hilbert or quadkey). Not sure what pattern this ZCurve is actually implementing tho.

That's absolutely correct. Our ZCurve is the same as quadkey. (Not the Hilbert.)
You will be able to use `ZCurve` for primary key.

In contrast, it doesn't make much sense for partition key, because you shouldn't consider partition key as index.

File: dbms/src/Functions/zCurveNormalized.cpp [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+                        if (range_right < left)
+                        {
+                            continue;
+                        }
+                    }
+                    append(ranges, range_left, range_right, nbits);
+                }
+            }
+            std::reverse(ranges.begin(), ranges.end());
+            for (auto& range : ranges)
```
Comment (4ertus2) [inline dbms/src/Functions/zCurveNormalized.cpp]: auto &

File: dbms/src/Functions/zCurveNormalized.cpp [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+
+        static std::vector<Segment> decodeRange(
+                ResultType left,
+                ResultType right,
+                const DataTypePtr & type,
+                size_t significant_digits)
+        {
+            size_t bits = sizeof(ResultType) << 3;
+            size_t type_bits = (type->getSizeOfValueInMemory() << 3);
+            size_t diff_bits = bits - type_bits;
```
Comment (4ertus2) [inline dbms/src/Functions/zCurveNormalized.cpp]: The second place we calculate it. Extract to own function
```
static size_t diffBits(const DataTypePtr & type);
```

File: dbms/src/Functions/zCurveNormalized.cpp [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+
+            std::vector<Segment> ranges;
+            ranges.emplace_back(left, right);
+
+            size_t lcp = __builtin_clzll(left ^ right);
+            ResultType clear_suf = (std::numeric_limits<ResultType>::max() << 1);
+
+            for (ResultType nbits = 1; nbits < type_bits; ++nbits, clear_suf <<= 1)
+            {
+                if (lcp + nbits >= bits)
```
Comment (4ertus2) [inline dbms/src/Functions/zCurveNormalized.cpp]: Comment is needed. What does it mean `lcp + nbits >= bits`

File: dbms/src/Functions/zCurveNormalized.cpp [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+                {
+                    ResultType suf_val = (left & clear_suf) | nbits;
+                    if (left <= suf_val && suf_val <= right)
+                    {
+                        append(ranges, suf_val, suf_val, nbits);
+                    }
+                }
+                else if (lcp + nbits + 1 == bits)
```
Comment (4ertus2) [inline dbms/src/Functions/zCurveNormalized.cpp]: Comment is needed

File: dbms/src/Functions/zCurveNormalized.cpp [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+                    {
+                        append(ranges, suf_max, suf_max, nbits);
+                    }
+                    if (left <= suf_min && suf_min <= right)
+                    {
+                        append(ranges, suf_min, suf_min, nbits);
+                    }
+                }
+                else
+                {
```
Comment (4ertus2) [inline dbms/src/Functions/zCurveNormalized.cpp]: Own function with good name is needed.

**Comment by 4ertus2** [comment]
It's error-prone to place both z-func and KeyConditions changes in one PR. It's hard to review it well. There's too much changes.

File: dbms/src/Functions/IFunction.h [hunk 6]
```diff
@@ -219,6 +223,17 @@ class IFunctionBase
...
         throw Exception("Function " + getName() + " has no information about its monotonicity.", ErrorCodes::NOT_IMPLEMENTED);
     }
+
+    /** This is used to work with the index in a sorted chunk of data and allows to efficiently
+      * use the index when the conditions are in the form `f(X) >= const`.
+      */
+    virtual bool isInvertible() const { return false; }
+
+    /// Computes the inverse image of given range of values in the form of a vector of parallelograms.
+    virtual bool invertRange(const Range& /*value range*/, size_t /*argument index*/, const DataTypes& /*argument types*/, RangeSet& /*result*/) const
```
Comment (stavrolia) [inline dbms/src/Functions/IFunction.h]: Range &, DataTypes &, RangeSet &

File: dbms/src/Functions/IFunction.h [hunk 7]
```diff
@@ -465,6 +480,13 @@ class DefaultFunction final : public IFunctionBase
 
     bool hasInformationAboutMonotonicity() const override { return function->hasInformationAboutMonotonicity(); }
 
+    bool isInvertible() const override { return function->isInvertible(); }
+
+    bool invertRange(const Range& value_range, size_t arg_index, const DataTypes& arg_types, RangeSet & result) const override
```
Comment (stavrolia) [inline dbms/src/Functions/IFunction.h]: Range &, DataTypes &

File: dbms/src/Functions/zCurve.cpp [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+
+
+namespace DB
+{
+
+    struct NameZCurve { static constexpr auto name = "zCurve"; };
+
+    ZCurveOpImpl::ResultType ZCurveOpImpl::decode(
+            ZCurveOpImpl::ResultType num,
+            const DB::DataTypePtr &type,
```
Comment (stavrolia) [inline dbms/src/Functions/zCurve.cpp]: & type

File: dbms/src/Functions/zCurve.cpp [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+            else if (num < NEG_INF)
+            {
+                num = NEG_INF;
+            }
+            num = RadixSortFloatTransform<UInt64>::backward(num);
+        }
+        return num;
+    }
+
+    void ZCurveOpImpl::encode(ZCurveOpImpl::ResultType &num, const DataTypePtr &type)
```
Comment (stavrolia) [inline dbms/src/Functions/zCurve.cpp]: & num, & type

File: dbms/src/Functions/zCurve.cpp [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+            num = RadixSortFloatTransform<UInt64>::forward(num);
+        }
+        // Shift the significant bits upwards
+        num <<= ((sizeof(ResultType) - type->getSizeOfValueInMemory()) << 3);
+    }
+
+    std::vector<std::pair<ZCurveOpImpl::ResultType, ZCurveOpImpl::ResultType>> ZCurveOpImpl::decodeRange(
+            ResultType left,
+            ResultType right,
+            const DataTypePtr &type,
```
Comment (stavrolia) [inline dbms/src/Functions/zCurve.cpp]: & type

File: dbms/src/Functions/zCurve.h [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+        using ResultType = UInt64;
+        /// Decode given z value range using the decode function. Always decodes into one range,
+        /// since the transformation is monotonous.
+        static std::vector<std::pair<ResultType, ResultType>> decodeRange(
+                ResultType left,
+                ResultType right,
+                const DataTypePtr & type,
+                size_t significant_digits);
+        /// Encodes different types monotonously into unsigned 64 bit integer representations.
+        static void encode(ResultType& num, const DataTypePtr & type);
```
Comment (stavrolia) [inline dbms/src/Functions/zCurve.h]: ResultType &

File: dbms/src/Functions/zCurveBase.h [hunk 12]
```diff
@@ -0,0 +1,342 @@
...
+            }
+            block.getByPosition(result).column = std::move(out);
+        }
+
+        bool isInvertible() const override
+        {
+            return true;
+        }
+
+        bool invertRange(const Range& value_range, size_t arg_index, const DataTypes& arg_types, RangeSet & result) const override
```
Comment (stavrolia) [inline dbms/src/Functions/zCurveBase.h]: Range &, DataTypes &

File: dbms/src/Functions/zCurveNormalized.cpp [hunk 13]
```diff
@@ -0,0 +1,146 @@
...
+
+namespace DB
+{
+
+    struct NameZCurveNormalized { static constexpr auto name = "zCurveN"; };
+
+    struct ZCurveNormalizedOpImpl
+    {
+        using ResultType = UInt64;
+        static void encode(ResultType& num, const DataTypePtr & type)
```
Comment (stavrolia) [inline dbms/src/Functions/zCurveNormalized.cpp]: ResultType &

File: dbms/src/Storages/MergeTree/FieldRange.cpp [hunk 14]
```diff
@@ -0,0 +1,300 @@
...
+            str << applyVisitor(FieldVisitorToString(), right) << (right_included ? ']' : ')');
+
+
+        return str.str();
+    }
+
+    bool Range::equals(const Field & lhs, const Field & rhs) { return applyVisitor(FieldVisitorAccurateEquals(), lhs, rhs); }
+    bool Range::less(const Field & lhs, const Field & rhs) { return applyVisitor(FieldVisitorAccurateLess(), lhs, rhs); }
+
+    RangeSet::RangeSet(const std::vector<DB::Range> &data): data(data)
```
Comment (stavrolia) [inline dbms/src/Storages/MergeTree/FieldRange.cpp]: & data

File: dbms/src/Storages/MergeTree/FieldRange.cpp [hunk 15]
```diff
@@ -0,0 +1,300 @@
...
+    {
+        normalize();
+    }
+    RangeSet::RangeSet(const DB::Range &range)
```
Comment (stavrolia) [inline dbms/src/Storages/MergeTree/FieldRange.cpp]: & range

File: dbms/src/Storages/MergeTree/FieldRange.cpp [hunk 16]
```diff
@@ -0,0 +1,300 @@
...
+            return !lhs.left_bounded ||
+            (rhs.left_bounded && (Range::less(lhs.left, rhs.left)
+            || (lhs.left_included && !rhs.left_included && Range::equals(lhs.left, rhs.left))
+            ));
+        });
+        std::vector<Range> normalized;
+        Field right_border;
+        bool right_bounded = false;
+        bool right_included = false;
+        for (const auto& range : data)
```
Comment (stavrolia) [inline dbms/src/Storages/MergeTree/FieldRange.cpp]: auto &

File: dbms/src/Storages/MergeTree/FieldRange.cpp [hunk 17]
```diff
@@ -0,0 +1,300 @@
...
+        return RangeSet(result);
+    }
+
+    /// Apply an invertible function to a set of ranges
+    std::optional<RangeSet> RangeSet::applyInvertibleFunction(
+            const FunctionBasePtr &func,
+            size_t arg_index)
+    {
+        RangeSet result;
+        for (const auto& range : data)
```
Comment (stavrolia) [inline dbms/src/Storages/MergeTree/FieldRange.cpp]: auto &

File: dbms/src/Storages/MergeTree/FieldRange.h [hunk 18]
```diff
@@ -0,0 +1,209 @@
...
+
+namespace DB
+{
+/** Range with open or closed ends; possibly unbounded.
+  */
+
+class Range
+{
+    public:
+        static bool equals(const Field &lhs, const Field &rhs);
```
Comment (stavrolia) [inline dbms/src/Storages/MergeTree/FieldRange.h]: & lhs, & rhs

File: dbms/src/Storages/MergeTree/FieldRange.h [hunk 19]
```diff
@@ -0,0 +1,209 @@
...
+
+        static bool less(const Field &lhs, const Field &rhs);
```
Comment (stavrolia) [inline dbms/src/Storages/MergeTree/FieldRange.h]: & lhs, & rhs

File: dbms/src/Storages/MergeTree/FieldRange.h [hunk 20]
```diff
@@ -0,0 +1,209 @@
...
+        bool left_bounded = false;        /// bounded at the left
+        bool right_bounded = false;       /// bounded at the right
+        bool left_included = false;       /// includes the left border, if any
+        bool right_included = false;      /// includes the right border, if any
+
+        /// The whole universum.
+        Range() {}
+
+        /// One point.
+        Range(const Field &point)
```
Comment (stavrolia) [inline dbms/src/Storages/MergeTree/FieldRange.h]: & point

File: dbms/src/Storages/MergeTree/FieldRange.h [hunk 21]
```diff
@@ -0,0 +1,209 @@
...
+                : left(point), right(point), left_bounded(true), right_bounded(true),
+                  left_included(true), right_included(true) {}
+
+        /// A bounded two-sided range.
+        Range(const Field &left_, bool left_included_, const Field &right_, bool right_included_)
```
Comment (stavrolia) [inline dbms/src/Storages/MergeTree/FieldRange.h]: & left_, & right_

File: dbms/src/Storages/MergeTree/FieldRange.h [hunk 22]
```diff
@@ -0,0 +1,209 @@
...
+                : left(left_), right(right_),
+                  left_bounded(true), right_bounded(true),
+                  left_included(left_included_), right_included(right_included_)
+        {
+            shrinkToIncludedIfPossible();
+        }
+
+        static Range createRightBounded(const Field &right_point, bool right_included)
```
Comment (stavrolia) [inline dbms/src/Storages/MergeTree/FieldRange.h]: & right_point

File: dbms/src/Storages/MergeTree/FieldRange.h [hunk 23]
```diff
@@ -0,0 +1,209 @@
...
+        {
+            Range r;
+            r.right = right_point;
+            r.right_bounded = true;
+            r.right_included = right_included;
+            r.shrinkToIncludedIfPossible();
+            return r;
+        }
+
+        static Range createLeftBounded(const Field &left_point, bool left_included)
```
Comment (stavrolia) [inline dbms/src/Storages/MergeTree/FieldRange.h]: & left_point

File: dbms/src/Storages/MergeTree/FieldRange.h [hunk 24]
```diff
@@ -0,0 +1,209 @@
...
+
+        bool empty() const
+        {
+            return left_bounded && right_bounded
+                   && (less(right, left)
+                       || ((!left_included || !right_included) && !less(left, right)));
+        }
+
+        /// x contained in the range
+        bool contains(const Field &x) const
```
Comment (stavrolia) [inline dbms/src/Storages/MergeTree/FieldRange.h]: & x

File: dbms/src/Storages/MergeTree/FieldRange.h [hunk 25]
```diff
@@ -0,0 +1,209 @@
...
+        {
+            return !leftThan(x) && !rightThan(x);
+        }
+
+        /// x is to the left
+        bool rightThan(const Field &x) const
```
Comment (stavrolia) [inline dbms/src/Storages/MergeTree/FieldRange.h]: & x

File: dbms/src/Storages/MergeTree/FieldRange.h [hunk 26]
```diff
@@ -0,0 +1,209 @@
...
+        {
+            return (left_bounded
+                    ? !(less(left, x) || (left_included && equals(x, left)))
+                    : false);
+        }
+
+        /// x is to the right
+        bool leftThan(const Field &x) const
```
Comment (stavrolia) [inline dbms/src/Storages/MergeTree/FieldRange.h]: & x

File: dbms/src/Storages/MergeTree/FieldRange.h [hunk 27]
```diff
@@ -0,0 +1,209 @@
...
+        {
+            return (right_bounded
+                    ? !(less(x, right) || (right_included && equals(x, right)))
+                    : false);
+        }
+
+        bool intersectsRange(const Range &r) const
```
Comment (stavrolia) [inline dbms/src/Storages/MergeTree/FieldRange.h]: & r

File: dbms/src/Storages/MergeTree/FieldRange.h [hunk 28]
```diff
@@ -0,0 +1,209 @@
...
+                && right_bounded
+                && (less(right, r.left)                          /// ...} {...
+                    || ((!right_included || !r.left_included)    /// ...) [... or ...] (...
+                        && equals(r.left, right))))
+                return false;
+
+            return true;
+        }
+
+        bool containsRange(const Range &r) const
```
Comment (stavrolia) [inline dbms/src/Storages/MergeTree/FieldRange.h]: & r

**Comment by stavrolia** [comment]
Please, fix other whitespaces in code too. You could use clang format (7th version is better) for it, our styleguide differs only in lambas.

File: dbms/src/Functions/zCurveBase.h [hunk 29]
```diff
@@ -0,0 +1,342 @@
...
+    template <typename Op, typename Name>
+    class FunctionZCurveBase : public IFunction
+    {
+    public:
+        static constexpr auto name = Name::name;
+        static FunctionPtr create(const Context & context)
+        {
+            return std::make_shared<FunctionZCurveBase>(context);
+        }
+        FunctionZCurveBase(const Context& /*context*/) {}
```
Comment (filimonov) [inline dbms/src/Functions/zCurveBase.h]:
```suggestion
        FunctionZCurveBase(const Context & /*context*/) {}
```

**Pull request closed without merging**